ArduPlane/systemid.cpp

```diff
@@ -0,0 +1,283 @@
+    // @User: Standard
+    AP_GROUPINFO("_XY_CTRL_MUL", 8, AP_SystemID, xy_control_mul, 0.25),
+    
+    AP_GROUPEND
+};
+
+AP_SystemID::AP_SystemID(void)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+// @LoggerMessage: SIDS
+// @Description: System ID settings
+// @Field: TimeUS: Time since system startup
+// @Field: Ax: The axis which is being excited
+// @Field: Mag: Magnitude of the chirp waveform
+// @Field: FSt: Frequency at the start of chirp
+// @Field: FSp: Frequency at the end of chirp
+// @Field: TFin: Time to reach maximum amplitude of chirp
+// @Field: TC: Time at constant frequency before chirp starts
+// @Field: TR: Time taken to complete chirp waveform
+// @Field: TFout: Time to reach zero amplitude after chirp finishes
+
+/*
+  start systemid
+ */
+void AP_SystemID::start()
+{
+    start_axis = axis;
+
+    // check if enabled
+    if (start_axis == AxisType::NONE) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "SystemID: No axis selected");
+        return;
+    }
+    if (!plane.control_mode->supports_systemid()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "SystemID: Not supported in mode %s", plane.control_mode->name());
+        return;
+    }
+    if (!hal.util->get_soft_armed()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "SystemID: must be armed");
+        return;
+    }
+    if (!plane.quadplane.enabled()) {
+        gcs().send_text(MAV_SEVERITY_WARNING, "SystemID: only for quadplane");
+        return;
+    }
+
+    attitude_offset_deg.zero();
+    throttle_offset = 0;
+
+    restore.att_bf_feedforward = plane.quadplane.attitude_control->get_bf_feedforward();
+
+    waveform_time = 0;
+    time_const_freq = 2.0 / frequency_start; // Two full cycles at the starting frequency
+
+    chirp_input.init(time_record, frequency_start, frequency_stop, time_fade_in, time_fade_out, time_const_freq);
+
+    gcs().send_text(MAV_SEVERITY_INFO, "SystemID Starting: axis=%d", (unsigned)axis);
+
+    AP::logger().WriteStreaming("SIDS", "TimeUS,Ax,Mag,FSt,FSp,TFin,TC,TR,TFout",
+                                "s--ssssss", "F--------", "QBfffffff",
+                                AP_HAL::micros64(),
+                                uint8_t(start_axis),
+                                waveform_magnitude.get(),
+                                frequency_start.get(),
+                                frequency_stop.get(),
+                                time_fade_in.get(),
+                                time_const_freq,
+                                time_record.get(),
+                                time_fade_out.get());
+
+    running = true;
+}
+
+/*
+  stop systemid
+ */
+void AP_SystemID::stop()
+{
+    if (running) {
+        running = false;
+        attitude_offset_deg.zero();
+        throttle_offset = 0;
+
+        auto *attitude_control = plane.quadplane.attitude_control;
+        attitude_control->bf_feedforward(restore.att_bf_feedforward);
+        attitude_control->rate_bf_roll_sysid(0);
```

Comment:
Setting these to zero is not necessary.  These are set to zero in the attitude controller at the end of every loop.
https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/quadplane.cpp#L1948

Reply:
this was paranoia to ensure that even a single loop is not incorrect

Reply:
changed to 1, and I still prefer to force it to 1, not rely on next loop